packages/contracts/contracts/Router.sol

```diff
@@ -0,0 +1,220 @@
+    uint256 _chainId,
+    address _routerSigner,
+    address _recipient,
+    address _owner
+  ) external onlyViaFactory {
+    transactionManager = ITransactionManager(_transactionManager);
+    chainId = _chainId;
+    routerSigner = _routerSigner;
+    recipient = _recipient;
+    transferOwnership(_owner);
+  }
+
+  function setRecipient(address _recipient) external onlyOwner {
+    recipient = _recipient;
+  }
+
+  function setSigner(address _routerSigner) external onlyOwner {
+    routerSigner = _routerSigner;
+  }
+
+  function addRelayerFee(uint256 amount, address assetId) external payable {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_ARF:002");
+
+    // Transfer funds to contract
+    // Validate correct amounts are transferred
+    if (LibAsset.isNativeAsset(assetId)) {
+      require(msg.value == amount, "#RC_ARF:005");
+    } else {
+      require(msg.value == 0, "#RC_ARF:006");
+      LibAsset.transferFromERC20(assetId, msg.sender, address(this), amount);
+    }
+
+    // Emit event
+    emit RelayerFeeAdded(assetId, amount, msg.sender);
+  }
+
+  function removeRelayerFee(uint256 amount, address assetId) external onlyOwner {
+    // Sanity check: nonzero amounts
+    require(amount > 0, "#RC_RRF:002");
+
+    // Transfer funds from contract
+    LibAsset.transferAsset(assetId, payable(recipient), amount);
+
+    // Emit event
+    emit RelayerFeeRemoved(assetId, amount, msg.sender);
+  }
+
+  function removeLiquidity(
+    uint256 amount,
+    address assetId,
+    bytes calldata signature
+  ) external {
+    if (msg.sender != routerSigner) {
+      SignedRemoveLiquidityData memory payload = SignedRemoveLiquidityData({
+        amount: amount,
+        assetId: assetId,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_RL:040");
+    }
+
+    return transactionManager.removeLiquidity(amount, assetId, payable(recipient));
+  }
+
+  function prepare(
+    ITransactionManager.PrepareArgs calldata args,
+    address routerRelayerFeeAsset,
+    uint256 routerRelayerFee,
+    bytes calldata signature
+  ) external payable returns (ITransactionManager.TransactionData memory) {
+    if (msg.sender != routerSigner) {
+      SignedPrepareData memory payload = SignedPrepareData({
+        args: args,
+        routerRelayerFeeAsset: routerRelayerFeeAsset,
+        routerRelayerFee: routerRelayerFee,
+        chainId: chainId
+      });
+
+      address recovered = recoverSignature(abi.encode(payload), signature);
+      require(recovered == routerSigner, "#RC_P:040");
+
+      // Send the relayer the fee
+      if (routerRelayerFee > 0) {
+        LibAsset.transferAsset(routerRelayerFeeAsset, payable(msg.sender), routerRelayerFee);
+      }
+    }
+
+    return
+      transactionManager.prepare{value: LibAsset.isNativeAsset(args.invariantData.sendingAssetId) ? msg.value : 0}(
+        args
+      );
+  }
```

Comment:
If the router sent any funds to the contract, it would need to be `receivingAssetId` , but all router liquidity is stored on chain already

should probably have some sanity checks around the chain id of this signer and the chain id of the transaction you're preparing

Reply:
Could also add more sanity checks around the router specified in the args and the calls

Reply:
shouldn't all of these checks be in the txmanager contract prepare method?

Reply:
yeah i guess they are, but doesn't mean im not a fan of excessive sanity checks in solidity

Reply:
I think this should be a super light wrapper contract and not do any of the extra validation on the solidity side tbh.